<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Connecting Bioconductor to other bioinformatics tools using Rcwl - Application in scRNA-seq data • RcwlscRNAseq</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Connecting Bioconductor to other bioinformatics tools using Rcwl - Application in scRNA-seq data">
<meta property="og:description" content="RcwlscRNAseq">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RcwlscRNAseq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/rcwl_scrnaseq.html">Workshop</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="rcwl_scrnaseq_files/header-attrs-2.7/header-attrs.js"></script><script src="rcwl_scrnaseq_files/accessible-code-block-0.0.1/empty-anchor.js"></script><script src="rcwl_scrnaseq_files/htmlwidgets-1.5.3/htmlwidgets.js"></script><script src="rcwl_scrnaseq_files/viz-1.8.2/viz.js"></script><link href="rcwl_scrnaseq_files/DiagrammeR-styles-0.2/styles.css" rel="stylesheet">
<script src="rcwl_scrnaseq_files/grViz-binding-1.0.6.1/grViz.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Connecting Bioconductor to other bioinformatics tools using Rcwl - Application in scRNA-seq data</h1>
            
      
      
      <div class="hidden name"><code>rcwl_scrnaseq.Rmd</code></div>

    </div>

    
    
<div id="instructors-names-and-contact-information" class="section level2">
<h2 class="hasAnchor">
<a href="#instructors-names-and-contact-information" class="anchor"></a>Instructor(s) name(s) and contact information</h2>
<ul>
<li>
<a href="https://github.com/liubuntu">Qian Liu</a> (<a href="mailto:Qian.Liu@roswellpark.org" class="email">Qian.Liu@roswellpark.org</a>)</li>
<li>
<a href="https://github.com/hubentu">Qiang Hu</a> (<a href="mailto:Qiang.Hu@roswellpark.org" class="email">Qiang.Hu@roswellpark.org</a>)</li>
</ul>
</div>
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>Here we demonstrate a case study for scRNA-seq data preprocessing using <code>RcwlPipeline</code> tools and pipeline.</p>
<p>10x Genomics has its own preprocessing pipeline <code>Cell Ranger</code> to process the scRNA-seq outputs it produces to perform the demultiplexing and quantification. However, it requires much configuration to run and is significantly slower than other mappers.</p>
<p>In this case study, <a href="https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md">STARsolo</a> is used for alignment and quantification, which produces a count matrix from FASTQ. <a href="https://bioconductor.org/packages/DropletUtils/">DropletUtils</a> is used for filtering raw gene-barcode matrix and removing empty droplets, which produces a high-quality count matrix with feature/cell annotation files saved in an <em>R</em> object of <code>SingleCellExperiment</code>. Before these, a one-time indexing step using <a href="https://hbctraining.github.io/Intro-to-rnaseq-hpc-O2/lessons/03_alignment.html">STARindex</a> is also included in this case study.</p>
</div>
<div id="install-packages" class="section level2">
<h2 class="hasAnchor">
<a href="#install-packages" class="anchor"></a>Install packages</h2>
<p>If this is the first time you use <code>Rcwl</code> or <code>RcwlPipelines</code>, you need to install the packages:</p>
<p>The package can be installed from Bioconductor (&gt;= 3.9):</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Rcwl"</span>, <span class="st">"RcwlPipelines"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Or use the most updated version from GitHub:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"rworkflow/Rcwl"</span>, <span class="st">"rworkflow/RcwlPipelines"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>For this workshop, we also need to install some other packages:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"git2r"</span>, <span class="st">"DropletUtils"</span>, <span class="st">"BiocParallel"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Load the packages:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">Rcwl</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">RcwlPipelines</span><span class="op">)</span></code></pre></div>
</div>
<div id="scrna-seq-data-source" class="section level2">
<h2 class="hasAnchor">
<a href="#scrna-seq-data-source" class="anchor"></a>scRNA-seq data source</h2>
<p>The scRNA-seq data source is the 1k PBMCs from 10x genomics (These source files are provided in the <a href="https://zenodo.org/record/3457880">Zenodo</a> data repository).</p>
<p>The dataset used in this tutorial are sub-sampled from the source files to contain only 15 cells instead of 1000. The data curation is for demo purpose only so that the execution of the Rcwl scRNA-seq preprocessing tools or pipeline in R can be completed within 1~2 minutes.</p>
<p>The "*.fastq" data was curated to only include reads on chromosome 21.</p>
<p>“subset15_demo_barcode.txt” contains known cell barcodes for mapping and only 15 barcodes are included.</p>
<p>“Homo_sapiens.GRCh37.75.21.gtf” contains the hg19 GTF file to annotate reads, which was curated on chromosome 21 only.</p>
<p>Data can be loaded from the dedicated <a href="https://github.com/rworkflow/testdata">GitHub repository</a> or the <a href="https://zenodo.org/record/4573881">Zenodo</a> data repository.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://docs.ropensci.org/git2r/">git2r</a></span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://docs.ropensci.org/git2r/reference/clone.html">clone</a></span><span class="op">(</span><span class="st">"https://github.com/rworkflow/testdata"</span>, <span class="st">"rcwl_data_supp"</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">path</span> <span class="op">&lt;-</span> <span class="st">"rcwl_data_supp"</span>  <span class="co">## source data</span>
<span class="fu"><a href="https://rdrr.io/r/base/list.files.html">dir</a></span><span class="op">(</span><span class="va">path</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "chr21.fa"                                         
## [2] "Homo_sapiens.GRCh37.75.21.gtf"                    
## [3] "subset15_chr21_pbmc_1k_v3_S1_L001_R1_001.fastq.gz"
## [4] "subset15_chr21_pbmc_1k_v3_S1_L001_R2_001.fastq.gz"
## [5] "subset15_chr21_pbmc_1k_v3_S1_L002_R1_001.fastq.gz"
## [6] "subset15_chr21_pbmc_1k_v3_S1_L002_R2_001.fastq.gz"
## [7] "subset15_demo_barcode.txt"</code></pre>
<p>Here we also create an output directory to save result files from running the tool/pipeline.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">outpath</span> <span class="op">&lt;-</span> <span class="st">"outdir"</span> 
<span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span><span class="op">(</span><span class="va">outpath</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<div id="load-scrna-seq-pre-processing-tools" class="section level3">
<h3 class="hasAnchor">
<a href="#load-scrna-seq-pre-processing-tools" class="anchor"></a>Load scRNA-seq pre-processing tools</h3>
<p>Three core functions: <code>cwlUpdate</code>, <code>cwlSearch</code> and <code>cwlLoad</code>from <code>RcwlPipelines</code> will be needed for updating, searching, and loading the needed tools or pipelines in <em>R</em>.</p>
<p>The <code>cwlUpdate</code> function syncs the current <code>Rcwl</code> recipes and returns a <code>cwlHub</code> object which contains the most updated <code>Rcwl</code> recipes. The <code>mcols()</code> function returns all related information about each available tool or pipeline. The recipes will be locally cached, so users don’t need to call <code>cwlUpdate</code> every time unless they want to use a tool/pipeline that is newly added to <code>RcwlPipelines</code>.</p>
<p>In this example, we are using the Bioc 3.13 which is the current devel version.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">atls</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/RcwlPipelines/man/cwlUpdate.html">cwlUpdate</a></span><span class="op">(</span>branch <span class="op">=</span> <span class="st">"dev"</span><span class="op">)</span> <span class="co">## sync the tools/pipelines. </span>
<span class="va">atls</span></code></pre></div>
<pre><code>## cwlHub with 140 records
## cache path:  ~/Library/Caches/Rcwl 
## # last modified date:  2021-02-22 
## # cwlSearch() to query scripts
## # cwlLoad('title') to load the script
## # additional mcols(): rid, rpath, Type, Container, mtime, ...
## 
##             title                      
##   BFC3498 | pl_alignMerge              
##   BFC3499 | pl_AnnPhaseVcf             
##   BFC3500 | pl_BaseRecal               
##   BFC3501 | pl_bwaAlign                
##   BFC3502 | pl_bwaMMRecal              
##   ...       ...                        
##   BFC3633 | tl_VarScan2                
##   BFC3634 | tl_vcf_expression_annotator
##   BFC3635 | tl_vcf_readcount_annotator 
##   BFC3636 | tl_vep                     
##   BFC3637 | tl_vt_decompose            
##           Command                                                            
##   BFC3498 bwaAlign+mergeBamDup                                               
##   BFC3499 VCFvep+dVCFcoverage+rVCFcoverage+VCFexpression+PhaseVcf            
##   BFC3500 BaseRecalibrator+ApplyBQSR+samtools_index+samtools_flagstat+samt...
##   BFC3501 bwa+sam2bam+sortBam+idxBam                                         
##   BFC3502 bwaAlign+mergeBamDup+BaseRecal                                     
##   ...     ...                                                                
##   BFC3633                                                                    
##   BFC3634 vcf-expression-annotator                                           
##   BFC3635 vcf-readcount-annotator                                            
##   BFC3636 vep                                                                
##   BFC3637 vt decompose</code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu">mcols</span><span class="op">(</span><span class="va">atls</span><span class="op">)</span><span class="op">$</span><span class="va">Type</span><span class="op">)</span></code></pre></div>
<pre><code>## 
## pipeline     tool 
##       26      113</code></pre>
<p><code>cwlSearch</code> is used to search for specific tools/pipelines of interest. Multiple keywords can be used for an internal search of “rname”, “rpath”, “fpath”, “Command” and “Containers” columns in the <code>mcols()</code>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tls</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/RcwlPipelines/man/cwlSearch.html">cwlSearch</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"STAR"</span>, <span class="st">"index"</span><span class="op">)</span><span class="op">)</span>
<span class="fu">mcols</span><span class="op">(</span><span class="va">tls</span><span class="op">)</span></code></pre></div>
<pre><code>## DataFrame with 2 rows and 14 columns
##           rid        rname         create_time         access_time
##   &lt;character&gt;  &lt;character&gt;         &lt;character&gt;         &lt;character&gt;
## 1     BFC3515 pl_rnaseq_Sf 2021-03-05 05:34:02 2021-03-05 05:34:02
## 2     BFC3618 tl_STARindex 2021-03-05 05:34:16 2021-03-05 05:34:16
##                    rpath       rtype                  fpath last_modified_time
##              &lt;character&gt; &lt;character&gt;            &lt;character&gt;          &lt;numeric&gt;
## 1 /Users/qi31566/Libra..       local /Users/qi31566/Libra..                 NA
## 2 /Users/qi31566/Libra..       local /Users/qi31566/Libra..                 NA
##          etag   expires        Type                Command
##   &lt;character&gt; &lt;numeric&gt; &lt;character&gt;            &lt;character&gt;
## 1          NA        NA    pipeline fastqc+STAR+sortBam+..
## 2          NA        NA        tool                   STAR
##                Container               mtime
##              &lt;character&gt;         &lt;character&gt;
## 1                     NA 2021-02-22 14:45:16
## 2 quay.io/biocontainer.. 2021-02-22 14:45:16</code></pre>
<p><code>cwlLoad</code> loads the <code>Rcwl</code> tool/pipeline into the <em>R</em> working environment. Here we load all the 3 tools that will be needed for the scRNA-seq data preprocessing. The recipes for developing these tools can be found in the <a href="https://github.com/rworkflow/RcwlRecipes/tree/master/Rcwl">GitHub repository</a> including <a href="https://github.com/rworkflow/RcwlRecipes/blob/master/Rcwl/tl_STARindex.R">tl_STARindex</a>, <a href="https://github.com/rworkflow/RcwlRecipes/blob/master/Rcwl/tl_STARsolo.R">tl_STARsolo</a>, and <a href="https://github.com/rworkflow/RcwlRecipes/blob/master/Rcwl/tl_DropletUtils.R">tl_DropletUtils</a>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">STARindex</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/RcwlPipelines/man/cwlLoad.html">cwlLoad</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/RcwlPipelines/man/cwlHub-methods.html">title</a></span><span class="op">(</span><span class="va">tls</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span>  <span class="co">## "tl_STARindex"</span>
<span class="co">## equivalent:</span>
<span class="co">## STARindex &lt;- cwlLoad(mcols(tls)$fpath[2])</span>
<span class="va">STARsolo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/RcwlPipelines/man/cwlLoad.html">cwlLoad</a></span><span class="op">(</span><span class="st">"tl_STARsolo"</span><span class="op">)</span>
<span class="va">DropletUtils</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/RcwlPipelines/man/cwlLoad.html">cwlLoad</a></span><span class="op">(</span><span class="st">"tl_DropletUtils"</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="scrna-seq-preprocessing" class="section level2">
<h2 class="hasAnchor">
<a href="#scrna-seq-preprocessing" class="anchor"></a>scRNA-seq preprocessing</h2>
<div id="indexing" class="section level3">
<h3 class="hasAnchor">
<a href="#indexing" class="anchor"></a>Indexing</h3>
<p>Before read alignment and quality control, a one-time genome indexing needs to be done. The command line using <code>STAR</code> will look like this:</p>
<pre><code>$ STAR --runMode genomeGenerate --runThreadN 4 --genomeDir STARindex 
--genomeFastaFiles chr21.fa --sjdbGTFfile Homo_sapiens.GRCh37.75.21.gtf</code></pre>
<p>We can equivalently index the genome using the <code>Rcwl</code> tool of <code>STARindex</code> within <em>R</em>, which was internally passed as cwl scripts, by only assigning values to the input parameters, and execute the cwl script using one of the execution functions, e.g., <code>runCWL</code> in the local computer. Then the output files are ready to pass as input to the next tool for single cell read alignment.</p>
<p>Assign values to the input parameters:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">STARindex</span><span class="op">$</span><span class="va">genomeFastaFiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">path</span>, <span class="st">"chr21.fa"</span><span class="op">)</span>
<span class="va">STARindex</span><span class="op">$</span><span class="va">sjdbGTFfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">path</span>, <span class="st">"Homo_sapiens.GRCh37.75.21.gtf"</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/Rcwl/man/runCWL.html">runCWL</a></span><span class="op">(</span>cwl <span class="op">=</span> <span class="va">STARindex</span>, outdir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">outpath</span>, <span class="st">"STARindex_output"</span><span class="op">)</span>, docker <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/list.files.html">dir</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">outpath</span>, <span class="st">"STARindex_output"</span><span class="op">)</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>  <span class="co">## output files</span></code></pre></div>
<pre><code>##  [1] "STARindex/chrLength.txt"            "STARindex/chrName.txt"             
##  [3] "STARindex/chrNameLength.txt"        "STARindex/chrStart.txt"            
##  [5] "STARindex/exonGeTrInfo.tab"         "STARindex/exonInfo.tab"            
##  [7] "STARindex/geneInfo.tab"             "STARindex/Genome"                  
##  [9] "STARindex/genomeParameters.txt"     "STARindex/Log.out"                 
## [11] "STARindex/SA"                       "STARindex/SAindex"                 
## [13] "STARindex/sjdbInfo.txt"             "STARindex/sjdbList.fromGTF.out.tab"
## [15] "STARindex/sjdbList.out.tab"         "STARindex/transcriptInfo.tab"</code></pre>
<p>Note that the <code>docker</code> argument in <code>runCWL</code> function takes 4 values:</p>
<ol style="list-style-type: decimal">
<li>TRUE (default, recommended), which automatically pulls docker images for the required command line tools.</li>
<li>FALSE, if users have already pre-installed all required command line tools.</li>
<li>“singularity” if the running environment doesn’t support docker but singularity.</li>
<li>“udocker” for the docker-like runtime without any administrator privileges.</li>
</ol>
</div>
<div id="alignment" class="section level3">
<h3 class="hasAnchor">
<a href="#alignment" class="anchor"></a>Alignment</h3>
<p>Assign values to input parameters:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cdna.fastq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">path</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="va">path</span>, pattern <span class="op">=</span> <span class="st">"_R2_"</span><span class="op">)</span><span class="op">)</span>
<span class="va">cb.fastq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">path</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="va">path</span>, pattern <span class="op">=</span> <span class="st">"_R1_"</span><span class="op">)</span><span class="op">)</span>
<span class="va">cblist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">path</span>, <span class="st">"subset15_demo_barcode.txt"</span><span class="op">)</span>
<span class="va">genomeDir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">outpath</span>, <span class="st">"STARindex_output/STARindex"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/Rcwl/man/cwlProcess-methods.html">inputs</a></span><span class="op">(</span><span class="va">STARsolo</span><span class="op">)</span></code></pre></div>
<pre><code>## inputs:
##   readFilesIn_cdna (File[]): --readFilesIn 
##   readFilesIn_cb (File[]):  
##   genomeDir (Directory): --genomeDir 
##   whiteList (File): --soloCBwhitelist 
##   soloType (string): --soloType CB_UMI_Simple
##   soloUMIlen (string): --soloUMIlen 12
##   runThreadN (int): --runThreadN 1</code></pre>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">STARsolo</span><span class="op">$</span><span class="va">readFilesIn_cdna</span> <span class="op">&lt;-</span> <span class="va">cdna.fastq</span>
<span class="va">STARsolo</span><span class="op">$</span><span class="va">readFilesIn_cb</span> <span class="op">&lt;-</span> <span class="va">cb.fastq</span>
<span class="va">STARsolo</span><span class="op">$</span><span class="va">whiteList</span> <span class="op">&lt;-</span> <span class="va">cblist</span>
<span class="va">STARsolo</span><span class="op">$</span><span class="va">genomeDir</span> <span class="op">&lt;-</span> <span class="va">genomeDir</span></code></pre></div>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/Rcwl/man/runCWL.html">runCWL</a></span><span class="op">(</span><span class="va">STARsolo</span>, outdir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">outpath</span>, <span class="st">"STARsolo_output"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/list.files.html">dir</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">outpath</span>, <span class="st">"STARsolo_output"</span><span class="op">)</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<pre><code>##  [1] "Aligned.out.sam"                     "Log.final.out"                      
##  [3] "Log.out"                             "Log.progress.out"                   
##  [5] "SJ.out.tab"                          "Solo.out/Barcodes.stats"            
##  [7] "Solo.out/Gene/Features.stats"        "Solo.out/Gene/filtered/barcodes.tsv"
##  [9] "Solo.out/Gene/filtered/features.tsv" "Solo.out/Gene/filtered/matrix.mtx"  
## [11] "Solo.out/Gene/raw/barcodes.tsv"      "Solo.out/Gene/raw/features.tsv"     
## [13] "Solo.out/Gene/raw/matrix.mtx"        "Solo.out/Gene/Summary.csv"          
## [15] "Solo.out/Gene/UMIperCellSorted.txt"</code></pre>
<p>The output files generated in the “STARsolo_output” folder can now be passed into the next tool for QC.</p>
</div>
<div id="count-filtering" class="section level3">
<h3 class="hasAnchor">
<a href="#count-filtering" class="anchor"></a>Count filtering</h3>
<p>To get a high-quality count matrix we apply the <code>DropletUtils</code> <em>Bioconductor</em> package, which will produce a filtered dataset that is more representative of the <code>Cell Ranger</code> pipeline.</p>
<p>Since CWL itself doesn’t support the integration of <em>R</em> packages or <em>R</em> function, this is a unique feature for <code>Rcwl</code>, where we can easily connect the upstream data preprocessing steps (previously based on command line tools) and the downstream data analysis steps (heavily done in <em>R/Bioconductor</em>).</p>
<p>The idea here is to put anything you need into a user-defined <em>R</em> function, with specified arguments for input and output files, then it’s ready to be wrapped as an <code>Rcwl</code> tools for execution.</p>
<p>For example, in wrapping the <em>Bioconductor</em> package <a href="https://bioconductor.org/packages/DropletUtils/">DropletUtils</a> functionalities, we wrote this <code>Rcwl</code> tool called <a href="https://github.com/rworkflow/RcwlRecipes/blob/master/Rcwl/tl_DropletUtils.R">tl_DropletUtils</a> with 3 major steps: 1) use the <code>read10xCounts</code> function to read the raw aligned files and convert into a <code>SingleCellExperiment</code> object. 2) calculate the barcode ranks and plotting. 3) calculate the empty droplets and plotting. We have also defined the tool output to collect the <code>SingleCellExperiment</code> Rdata file and diagnostic figures pdf file.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">DropletUtils</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/Rcwl/man/cwlProcess-methods.html">inputs</a></span><span class="op">(</span><span class="va">DropletUtils</span><span class="op">)</span></code></pre></div>
<pre><code>## inputs:
##   dirname (Directory): dir.name= 
##   lower (int): lower= 100
##   df (int): df= 20</code></pre>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">DropletUtils</span><span class="op">$</span><span class="va">dirname</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">outpath</span>, <span class="st">"STARsolo_output/Solo.out"</span><span class="op">)</span>
<span class="va">DropletUtils</span><span class="op">$</span><span class="va">lower</span> <span class="op">&lt;-</span> <span class="fl">100</span>
<span class="va">DropletUtils</span><span class="op">$</span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fl">5</span></code></pre></div>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/Rcwl/man/runCWL.html">runCWL</a></span><span class="op">(</span><span class="va">DropletUtils</span>, outdir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">outpath</span>, <span class="st">"dropletUtils_output"</span><span class="op">)</span>, showLog <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/list.files.html">dir</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">outpath</span>, <span class="st">"dropletUtils_output"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "diagnostics.pdf"  "sce_filtered.rds"</code></pre>
<p>Now that we get 2 output files:</p>
<ol style="list-style-type: decimal">
<li>The pdf file with 2 diagnostic figures: Barcode ranks, and empty droplets. For details regarding interpretation of each diagnostic figure, please refer to the <a href="https://bioconductor.org/packages/release/bioc/vignettes/DropletUtils/inst/doc/DropletUtils.html">DropletUtils vignette</a>.</li>
</ol>
<p><img src="diag1.png" width="60%"><img src="diag2.png" width="60%"></p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">Biobase</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/openPDF.html">openPDF</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">outpath</span>, <span class="st">"dropletUtils_output/diagnostics.pdf"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>The <code>SingleCellExperiment</code> object which has filtered out unqualified cells and is analysis ready.</li>
</ol>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">outpath</span>, <span class="st">"dropletUtils_output/sce_filtered.rds"</span><span class="op">)</span><span class="op">)</span>
<span class="va">sce</span></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 736 10 
## metadata(1): Samples
## assays(1): counts
## rownames(736): ENSG00000238411 ENSG00000264462 ... ENSG00000230982
##   ENSG00000212932
## rowData names(3): ID Symbol Type
## colnames: NULL
## colData names(2): Sample Barcode
## reducedDimNames(0):
## altExpNames(0):</code></pre>
</div>
<div id="scrna-seq-preprocessing-pipeline" class="section level3">
<h3 class="hasAnchor">
<a href="#scrna-seq-preprocessing-pipeline" class="anchor"></a>scRNA-seq preprocessing pipeline</h3>
<p>Alternatively and more easily, we can use the pipeline called <code>pl_STARsoloDropletUtils</code> (recipe <a href="https://github.com/rworkflow/RcwlRecipes/blob/master/Rcwl/pl_STARsoloDropletUtils.R">here</a>) for the scRNA-seq data preprocessing. This pipeline integrated the <code>STARsolo</code> and <code>DropletUtils</code> for a streamlined preprocessing analysis within <em>R</em>. It is included in <code>RcwlPipelines</code> and is ready to be customized for your own research. Pipelines can be visualized by the <code>plotCWL</code> function.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">STARsoloDropletUtils</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/RcwlPipelines/man/cwlLoad.html">cwlLoad</a></span><span class="op">(</span><span class="st">"pl_STARsoloDropletUtils"</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/Rcwl/man/plotCWL.html">plotCWL</a></span><span class="op">(</span><span class="va">STARsoloDropletUtils</span><span class="op">)</span></code></pre></div>
<div id="htmlwidget-cbbb00076028495554ac" style="width:700px;height:432.632880098888px;" class="grViz html-widget"></div>
<script type="application/json" data-for="htmlwidget-cbbb00076028495554ac">{"x":{"diagram":"digraph {\n\ngraph [layout = \"neato\",\n       outputorder = \"edgesfirst\",\n       bgcolor = \"white\"]\n\nnode [fontname = \"Helvetica\",\n      fontsize = \"10\",\n      shape = \"circle\",\n      fixedsize = \"true\",\n      width = \"0.5\",\n      style = \"filled\",\n      fillcolor = \"aliceblue\",\n      color = \"gray70\",\n      fontcolor = \"gray50\"]\n\nedge [fontname = \"Helvetica\",\n     fontsize = \"8\",\n     len = \"1.5\",\n     color = \"gray80\",\n     arrowsize = \"0.5\"]\n\n  \"1\" [label = \"fastq_cdna\", shape = \"rectangle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"0,5!\"] \n  \"2\" [label = \"fastq_cb\", shape = \"rectangle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"1,5!\"] \n  \"3\" [label = \"genomeDir\", shape = \"rectangle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"2,5!\"] \n  \"4\" [label = \"whiteList\", shape = \"rectangle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"3,5!\"] \n  \"5\" [label = \"runThreadN\", shape = \"rectangle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"4,5!\"] \n  \"6\" [label = \"sam\", shape = \"circle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"1.5,3!\"] \n  \"7\" [label = \"Solo\", shape = \"circle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"2.5,3!\"] \n  \"8\" [label = \"sce\", shape = \"circle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"2,1!\"] \n  \"9\" [label = \"plots\", shape = \"circle\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"3,1!\"] \n  \"10\" [label = \"STARsolo\", shape = \"diamond\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"2,4!\"] \n  \"11\" [label = \"DropletUtils\", shape = \"diamond\", fillcolor = \"#F0F8FF\", fontcolor = \"#000000\", pos = \"2.5,2!\"] \n  \"1\"->\"10\" \n  \"2\"->\"10\" \n  \"3\"->\"10\" \n  \"4\"->\"10\" \n  \"5\"->\"10\" \n  \"10\"->\"6\" \n  \"10\"->\"7\" \n  \"7\"->\"11\" \n  \"11\"->\"9\" \n  \"11\"->\"8\" \n}","config":{"engine":"dot","options":null}},"evals":[],"jsHooks":[]}</script><!-- ```{r fig2, fig.show = "workflow", out.width = "100%", echo = FALSE} --><!-- knitr::include_graphics("workflow.png") --><!-- ``` --><p>For <code>Rcwl</code> pipelines, we only need to assign input values for the whole pipeline, not individual tools involved. The input and output between each step are pre-defined in the pipeline to ensure a smooth passing.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/Rcwl/man/cwlProcess-methods.html">inputs</a></span><span class="op">(</span><span class="va">STARsoloDropletUtils</span><span class="op">)</span></code></pre></div>
<pre><code>## inputs:
##   fastq_cdna (File[]):  
##   fastq_cb (File[]):  
##   genomeDir (Directory):  
##   whiteList (File):  
##   runThreadN (int):</code></pre>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">STARsoloDropletUtils</span><span class="op">$</span><span class="va">fastq_cdna</span> <span class="op">&lt;-</span> <span class="va">cdna.fastq</span>
<span class="va">STARsoloDropletUtils</span><span class="op">$</span><span class="va">fastq_cb</span> <span class="op">&lt;-</span> <span class="va">cb.fastq</span>
<span class="va">STARsoloDropletUtils</span><span class="op">$</span><span class="va">genomeDir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">outpath</span>, <span class="st">"STARindex_output/STARindex"</span><span class="op">)</span>
<span class="va">STARsoloDropletUtils</span><span class="op">$</span><span class="va">whiteList</span> <span class="op">&lt;-</span> <span class="va">cblist</span>
<span class="va">STARsoloDropletUtils</span><span class="op">$</span><span class="va">runThreadN</span> <span class="op">&lt;-</span> <span class="fl">1</span></code></pre></div>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/Rcwl/man/runCWL.html">runCWL</a></span><span class="op">(</span><span class="va">STARsoloDropletUtils</span>, outdir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">outpath</span>, <span class="st">"scpipeline_output"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>The overall output of the pipeline was pre-defined to glob the important files from separate steps.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/Rcwl/man/cwlProcess-methods.html">outputs</a></span><span class="op">(</span><span class="va">STARsoloDropletUtils</span><span class="op">)</span></code></pre></div>
<pre><code>## outputs:
## sam:
##   type: File
##   outputSource: STARsolo/outAlign
## Solo:
##   type: Directory
##   outputSource: STARsolo/Solo
## sce:
##   type: File
##   outputSource: DropletUtils/outsce
## plots:
##   type: File
##   outputSource: DropletUtils/plots</code></pre>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/list.files.html">dir</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">outpath</span>, <span class="st">"scpipeline_output"</span><span class="op">)</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<pre><code>##  [1] "Aligned.out.sam"                     "diagnostics.pdf"                    
##  [3] "sce_filtered.rds"                    "Solo.out/Barcodes.stats"            
##  [5] "Solo.out/Gene/Features.stats"        "Solo.out/Gene/filtered/barcodes.tsv"
##  [7] "Solo.out/Gene/filtered/features.tsv" "Solo.out/Gene/filtered/matrix.mtx"  
##  [9] "Solo.out/Gene/raw/barcodes.tsv"      "Solo.out/Gene/raw/features.tsv"     
## [11] "Solo.out/Gene/raw/matrix.mtx"        "Solo.out/Gene/Summary.csv"          
## [13] "Solo.out/Gene/UMIperCellSorted.txt"</code></pre>
</div>
<div id="submit-parallel-jobs-for-starsolo" class="section level3">
<h3 class="hasAnchor">
<a href="#submit-parallel-jobs-for-starsolo" class="anchor"></a>Submit parallel jobs for <code>STARsolo</code>
</h3>
<p>Powered by <code>BiocParallel</code>,<code>Rcwl</code> supports parallel job running for multiple samples using the <code>runCWLBatch</code> function. The following example demonstrates how to do the parallel alignment for the 2 samples using the <code>STARsolo</code> tool.</p>
<p>The <code>BPPARAM</code> argument in <code><a href="https://rdrr.io/pkg/Rcwl/man/runCWLBatch.html">runCWLBatch()</a></code> defines the parallel parameters. It can be defined by <code><a href="https://rdrr.io/pkg/BiocParallel/man/BatchtoolsParam-class.html">BiocParallel::BatchtoolsParam</a></code> function, where the <code>cluster</code> argument takes different values for different cluster job manager, such as “multicore”, “sge” and “slurm”. More details about available options can be checked by <code><a href="https://rdrr.io/pkg/BiocParallel/man/BatchtoolsParam-class.html">?BiocParallel::BatchtoolsParam</a></code>.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel">BiocParallel</a></span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bpparam</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/BatchtoolsParam-class.html">BatchtoolsParam</a></span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">2</span>, cluster <span class="op">=</span> <span class="st">"sge"</span>,
                           template <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/BatchtoolsParam-class.html">batchtoolsTemplate</a></span><span class="op">(</span><span class="st">"sge"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>In the following example, we are using “multicore” for the parallel running.</p>
<p>The <code>inputList</code> argument is required to be a list of input parameter values for samples that are to be computed parallelly. <strong>NOTE</strong> that the names of the list must be consistent with the ids of input parameters. In this example, the names are <code>readFilesIn_cdna</code> and <code>readFilesIn_cb</code>.</p>
<p>The <code>paramList</code> argument is required to be a list of input parameter values that are to be shared for all parallelly running samples.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bpparam</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/BatchtoolsParam-class.html">BatchtoolsParam</a></span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">2</span>, cluster <span class="op">=</span> <span class="st">"multicore"</span><span class="op">)</span>
<span class="va">input_lst</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>readFilesIn_cdna <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>
                      sample1 <span class="op">=</span> <span class="va">cdna.fastq</span>,
                      sample2 <span class="op">=</span> <span class="va">cdna.fastq</span><span class="op">)</span>,
                  readFilesIn_cb <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>
                      sample1 <span class="op">=</span> <span class="va">cb.fastq</span>,
                      sample2 <span class="op">=</span> <span class="va">cb.fastq</span><span class="op">)</span>
                  <span class="op">)</span>
<span class="va">param_lst</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>whiteList <span class="op">=</span> <span class="va">cblist</span>,
                  genomeDir <span class="op">=</span> <span class="va">genomeDir</span>,
                  runThreadN <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Rcwl/man/runCWLBatch.html">runCWLBatch</a></span><span class="op">(</span>cwl <span class="op">=</span> <span class="va">STARsolo</span>,
                   outdir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">outpath</span>, <span class="st">"STARsolo_batch_output"</span><span class="op">)</span>,
                   inputList <span class="op">=</span> <span class="va">input_lst</span>, paramList <span class="op">=</span> <span class="va">param_lst</span>, 
                   BPPARAM <span class="op">=</span> <span class="va">bpparam</span><span class="op">)</span></code></pre></div>
<p>The results are saved in separate folders for each parallel sample.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/list.files.html">dir</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">outpath</span>, <span class="st">"STARsolo_batch_output"</span><span class="op">)</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>  <span class="co">## output files</span></code></pre></div>
<pre><code>##  [1] "sample1/Aligned.out.sam"                    
##  [2] "sample1/Log.final.out"                      
##  [3] "sample1/Log.out"                            
##  [4] "sample1/Log.progress.out"                   
##  [5] "sample1/SJ.out.tab"                         
##  [6] "sample1/Solo.out/Barcodes.stats"            
##  [7] "sample1/Solo.out/Gene/Features.stats"       
##  [8] "sample1/Solo.out/Gene/filtered/barcodes.tsv"
##  [9] "sample1/Solo.out/Gene/filtered/features.tsv"
## [10] "sample1/Solo.out/Gene/filtered/matrix.mtx"  
## [11] "sample1/Solo.out/Gene/raw/barcodes.tsv"     
## [12] "sample1/Solo.out/Gene/raw/features.tsv"     
## [13] "sample1/Solo.out/Gene/raw/matrix.mtx"       
## [14] "sample1/Solo.out/Gene/Summary.csv"          
## [15] "sample1/Solo.out/Gene/UMIperCellSorted.txt" 
## [16] "sample2/Aligned.out.sam"                    
## [17] "sample2/Log.final.out"                      
## [18] "sample2/Log.out"                            
## [19] "sample2/Log.progress.out"                   
## [20] "sample2/SJ.out.tab"                         
## [21] "sample2/Solo.out/Barcodes.stats"            
## [22] "sample2/Solo.out/Gene/Features.stats"       
## [23] "sample2/Solo.out/Gene/filtered/barcodes.tsv"
## [24] "sample2/Solo.out/Gene/filtered/features.tsv"
## [25] "sample2/Solo.out/Gene/filtered/matrix.mtx"  
## [26] "sample2/Solo.out/Gene/raw/barcodes.tsv"     
## [27] "sample2/Solo.out/Gene/raw/features.tsv"     
## [28] "sample2/Solo.out/Gene/raw/matrix.mtx"       
## [29] "sample2/Solo.out/Gene/Summary.csv"          
## [30] "sample2/Solo.out/Gene/UMIperCellSorted.txt"</code></pre>
</div>
<div id="shiny-interface" class="section level3">
<h3 class="hasAnchor">
<a href="#shiny-interface" class="anchor"></a>Shiny interface</h3>
<p><code><a href="https://rdrr.io/pkg/Rcwl/man/cwlShiny.html">cwlShiny()</a></code> opens a user-friendly shiny interface for running any <code>Rcwl</code> tools or pipelines. By default, users need to put in the absolute file path for each input parameter. <strong>NOTE</strong>, multiple file paths need to be separated by colon. Click the <code>run</code> button, it will start running and return the output file paths under <code>Output</code> tag. Users can also check the <code>Command</code> and <code>Log</code> in the shiny interface page.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/Rcwl/man/cwlShiny.html">cwlShiny</a></span><span class="op">(</span><span class="va">STARsolo</span><span class="op">)</span></code></pre></div>
<p><img src="cwlShiny_STARsolo.png" width="100%"></p>
</div>
</div>
<div id="sessioninfo" class="section level2">
<h2 class="hasAnchor">
<a href="#sessioninfo" class="anchor"></a>SessionInfo</h2>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## R version 4.0.3 (2020-10-10)
## Platform: x86_64-apple-darwin13.4.0 (64-bit)
## Running under: macOS Catalina 10.15.7
## 
## Matrix products: default
## BLAS/LAPACK: /Users/qi31566/miniconda3/envs/r-base/lib/libopenblasp-r0.3.12.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] BiocParallel_1.24.1         DropletUtils_1.10.3        
##  [3] SingleCellExperiment_1.12.0 SummarizedExperiment_1.20.0
##  [5] Biobase_2.50.0              GenomicRanges_1.42.0       
##  [7] GenomeInfoDb_1.26.2         IRanges_2.24.1             
##  [9] MatrixGenerics_1.2.1        matrixStats_0.58.0         
## [11] git2r_0.28.0                RcwlPipelines_1.7.7        
## [13] BiocFileCache_1.14.0        dbplyr_2.1.0               
## [15] Rcwl_1.7.12                 S4Vectors_0.28.1           
## [17] BiocGenerics_0.36.0         yaml_2.2.1                 
## 
## loaded via a namespace (and not attached):
##   [1] ellipsis_0.3.1            rprojroot_2.0.2          
##   [3] scuttle_1.0.4             XVector_0.30.0           
##   [5] fs_1.5.0                  rstudioapi_0.13          
##   [7] bit64_4.0.5               fansi_0.4.2              
##   [9] sparseMatrixStats_1.2.1   codetools_0.2-18         
##  [11] R.methodsS3_1.8.1         cachem_1.0.4             
##  [13] knitr_1.31                jsonlite_1.7.2           
##  [15] R.oo_1.24.0               shiny_1.6.0              
##  [17] HDF5Array_1.18.1          DiagrammeR_1.0.6.1       
##  [19] compiler_4.0.3            httr_1.4.2               
##  [21] dqrng_0.2.1               basilisk_1.2.1           
##  [23] backports_1.2.1           assertthat_0.2.1         
##  [25] Matrix_1.3-2              fastmap_1.1.0            
##  [27] limma_3.46.0              later_1.1.0.1            
##  [29] visNetwork_2.0.9          htmltools_0.5.1.1        
##  [31] prettyunits_1.1.1         tools_4.0.3              
##  [33] igraph_1.2.6              glue_1.4.2               
##  [35] GenomeInfoDbData_1.2.4    dplyr_1.0.4              
##  [37] batchtools_0.9.15         rappdirs_0.3.3           
##  [39] Rcpp_1.0.6                jquerylib_0.1.3          
##  [41] pkgdown_1.6.1             vctrs_0.3.6              
##  [43] rhdf5filters_1.2.0        DelayedMatrixStats_1.12.3
##  [45] xfun_0.21                 stringr_1.4.0            
##  [47] beachmat_2.6.4            mime_0.10                
##  [49] lifecycle_1.0.0           edgeR_3.32.1             
##  [51] zlibbioc_1.36.0           basilisk.utils_1.2.2     
##  [53] ragg_1.1.0                hms_1.0.0                
##  [55] promises_1.2.0.1          rhdf5_2.34.0             
##  [57] RColorBrewer_1.1-2        curl_4.3                 
##  [59] memoise_2.0.0             reticulate_1.18          
##  [61] sass_0.3.1                stringi_1.5.3            
##  [63] RSQLite_2.2.3             highr_0.8                
##  [65] desc_1.2.0                checkmate_2.0.0          
##  [67] filelock_1.0.2            rlang_0.4.10             
##  [69] pkgconfig_2.0.3           systemfonts_1.0.1        
##  [71] bitops_1.0-6              evaluate_0.14            
##  [73] lattice_0.20-41           purrr_0.3.4              
##  [75] Rhdf5lib_1.12.1           htmlwidgets_1.5.3        
##  [77] bit_4.0.4                 tidyselect_1.1.0         
##  [79] magrittr_2.0.1            R6_2.5.0                 
##  [81] generics_0.1.0            base64url_1.4            
##  [83] DelayedArray_0.16.1       DBI_1.1.1                
##  [85] pillar_1.5.0              withr_2.4.1              
##  [87] RCurl_1.98-1.2            tibble_3.0.6             
##  [89] crayon_1.4.1              utf8_1.1.4               
##  [91] rmarkdown_2.7             progress_1.2.2           
##  [93] locfit_1.5-9.4            grid_4.0.3               
##  [95] data.table_1.14.0         blob_1.2.1               
##  [97] digest_0.6.27             xtable_1.8-4             
##  [99] tidyr_1.1.2               httpuv_1.5.5             
## [101] brew_1.0-6                R.utils_2.10.1           
## [103] textshaping_0.3.1         bslib_0.2.4</code></pre>
</div>
<div id="links" class="section level2">
<h2 class="hasAnchor">
<a href="#links" class="anchor"></a>Links</h2>
<ol style="list-style-type: decimal">
<li>
<a href="https://github.com/rworkflow/Rcwl_scRNAseq">GitHub repository</a> for the workshop materials and instructions.</li>
<li>
<a href="https://youtu.be/5jkJAfkjFGo">YouTube link</a> for the Rcwl workshop on BioCAsia2020.</li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Qian Liu, Qiang Hu.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
